### PR TITLE
Add a convention about z_index

### LINKED ISSUE #9606: [discuss] Come up with a convention for HUD element z_index
This is a discussion issue.

In 5.2.0, `z_index` was added to HUD elements to determine the rendering order. As many different mods are manipulating the HUD, they will use `z_index`.

However, there is no convention whatsoever as to what numbers to use. I feel we should come up with *some* convention (even a simple one). Otherwise, the mods will just spread the `z_index`es all over the place, making it useless. Therefore, I suggest to discuss a convention, and when the discussion is complete, put it into `lua_api.txt` or at least the Dev Wiki as official recommendation.

My first rough ideas:

* -40: Graphical effects, such as vignette
* -30: Name tags, waypoints
* -20: Wieldhand
* -10: Things you wear on your face or that are blocking your view, such as a mask or a bloddy screen
* 0: Default `z_index`. For the builtin HUD elements and anything that can be considered to be part of the “true” HUD: Minimap, hotbar, builtin statbars, crosshair
* 10: Temporary text messages or notification icons
* 100: Full-screen effects such are full-black screen or a credits screen. Basically anything that must cover the full screen and replace the entire HUD

These numbers have intentionally have gaps so it's easy to put in new stuff in-between. It's also neccessary for complex HUDs that need multiple elements at once for the same thing.

Do you agree? Do you have your own ideas for a convention?

### PR DESCRIPTION
**Pull request by Wuzzy2**
Fixes #9606.

### CODE REVIEW

**Comment by sfan5** [comment]
The code should also be modified to respect that convention, no?

**Comment by Wuzzy2** [comment]
Ummmmmm, sorry, I have no idea where to look. :-/
Sorry.

**Comment by rubenwardy** [comment]
It'll be in hud.cpp, not sure if the default stuff supports z-layers though

**Comment by SmallJoker** [comment]
HUD elements can be used for various purposes. What would need updating are the statbars (builtin/game) and the inventory hotlist (probably builtin too).

**Comment by Wuzzy2** [comment]
I believe the engine itself does not use `z_index` so far, except for the builtin statbars maybe (which are `z_index=0`, so no change neccessary).
 I don't think lacking usage of the engine is a good enough reason to block this PR. If we wait too long with adopting at least some kind of convention, things will get messy.

File: doc/lua_api.txt [hunk 1]
```diff
@@ -1310,7 +1310,24 @@ factor!
...
+*  -200: Wieldhand
+*  -100: Things you wear on your face or stuff that is blocking your view,
+         such as a mask
+*     0: Default. For builtin HUD elements and anything that is
+         considered to be the “true” HUD:
+         Minimap, hotbar, builtin statbars, crosshair
+*   100: Temporary text messages or notification icons
+*  1000: Full-screen effects such as full-black screen or credits.
+         Basically anything that must cover the entire screen and
+         replace the entire HUD
```
Comment (SmallJoker) [inline doc/lua_api.txt:1326]:
```suggestion
         This includes effects that cover the entire screen.
```

File: doc/lua_api.txt [hunk 2]
```diff
@@ -1310,7 +1310,24 @@ factor!
...
 Lower z-index elements are displayed behind higher z-index elements. Elements
 with same z-index are displayed in an arbitrary order. Default 0.
-Supports negative values.
+Supports negative values. By convention, the following values are recommended:
+
+*  -400: Graphical effects, such as vignette
+*  -300: Name tags, waypoints
+*  -200: Wieldhand
+*  -100: Things you wear on your face or stuff that is blocking your view,
+         such as a mask
```
Comment (SmallJoker) [inline doc/lua_api.txt:1319]: In general, I think direct speech should be avoided (-> `you`).
For example: `Elements that block the player's view, e.g. masks.`

File: doc/lua_api.txt [hunk 3]
```diff
@@ -1310,7 +1310,24 @@ factor!
...
+*     0: Default. For builtin HUD elements and anything that is
+         considered to be the “true” HUD:
+         Minimap, hotbar, builtin statbars, crosshair
```
Comment (SmallJoker) [inline doc/lua_api.txt:1322]: How about:
`Standard in-game HUD elements that are usually visible: Minimap, hotbar...`

**Comment by Wuzzy2** [comment]
Hopefully it's better now. :)

File: doc/lua_api.txt [hunk 4]
```diff
@@ -1310,7 +1310,21 @@ factor!
...
+Supports negative values. By convention, the following values are recommended:
+
+*  -400: Graphical effects, such as vignette
+*  -300: Name tags, waypoints
+*  -200: Wieldhand
+*  -100: Things that block the player's view, e.g. masks
+*     0: Default. For standard in-game HUD elements like crosshair, hotbar,
+         minimap, builtin statbars, etc.
+*   100: Temporary text messages or notification icons
+*  1000: Full-screen effects such as full-black screen or credits.
```
Comment (appgurueu) [inline doc/lua_api.txt:1322]: How about a smaller number?

**Comment by Wuzzy2** [comment]
The large numbers are intentionally large so we can cram more stuff in between for those cases we haven't thought of yet. Was suggested by sfan5, and I agree.
`z_index` is a signed 16-bit integer, so we still have plenty of space left anyway.

**Comment by nerzhul** [comment]
Good point to prepare convention around z-index :)

**Pull request merged**